File: src/characters/poisoner.ml
open! Core
open! Import

include Character.Make (struct
    type t =
      { kind : Kind.t
      ; alignment : Alignment.t
      }

    let id = "poisoner"
    let name = "Poisoner"
    let t = { kind = Kind.Minion; alignment = Alignment.Evil }
    let kind { kind; _ } = kind
    let alignment { alignment; _ } = alignment
  end)

let night_action ~player_id:pid ~night:_ =
  Some
    (if_alive
       pid
       (let%bind.Botc_exec state = Botc_exec.get_state
        and () = Botc_exec.wake pid in
        let%bind.Botc_exec target =
          Botc_exec.ask pid "Who do you want to poison?" (alive_except state pid)
        in
        let%bind.Botc_exec () =
          Botc_exec.modify_state (fun s -> Game_state.set_poisoned s target)
        in
        Botc_exec.sleep pid))
;;

let day_action ~player_id:_ = None
let on_setup ~player_id:_ = None
let on_nominated ~player_id:_ ~nominator:_ = None
let on_executed ~player_id:_ = None
let on_night_kill ~player_id:_ = None

File: src/characters/poisoner.rs
//! Poisoner (Minion): each night, choose a player; they are poisoned until you
//! act again or die.

use crate::ability::Ability;
use crate::ctx::Ctx;
use crate::event::DeathSource;
use crate::ids::PlayerId;
use crate::role::{Alignment, CharacterInfo, Kind};
use crate::token::Token;

character!(
    Poisoner,
    "poisoner",
    "Poisoner",
    Kind::Minion,
    Alignment::Evil
);

impl Ability for Poisoner {
    fn info(&self) -> CharacterInfo {
        Self::INFO
    }

    fn on_night(&self, ctx: &mut Ctx, me: PlayerId, _night: u32) {
        if !ctx.grim.is_alive(me) {
            return;
        }
        ctx.wake(me);
        let targets = ctx.alive_others(me);
        let target = ctx.ask(me, "Who do you want to poison?", &targets);
        // Move the single poison token from wherever it was to the new target.
        ctx.grim.clear_token_kind_everywhere(Token::Poisoned);
        ctx.grim.add_token(target, Token::Poisoned);
        ctx.sleep(me);
    }

    fn on_any_death(&self, ctx: &mut Ctx, me: PlayerId, dead: PlayerId, _source: DeathSource) {
        // When the Poisoner dies, their poison wears off.
        if me == dead {
            ctx.grim.clear_token_kind_everywhere(Token::Poisoned);
        }
    }
}
